flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -0,0 +1,290 @@
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext, IcebergEnumeratorState enumState) {
+    return createEnumerator(enumContext, enumState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergSourceSplit> getSplitSerializer() {
+    return IcebergSourceSplitSerializer.INSTANCE;
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergEnumeratorState> getEnumeratorCheckpointSerializer() {
+    return IcebergEnumeratorStateSerializer.INSTANCE;
+  }
+
+  private SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext,
+      @Nullable IcebergEnumeratorState enumState) {
+    Table table = loadTable(tableLoader);
+    SplitAssigner assigner;
+    if (enumState == null) {
+      assigner = assignerFactory.createAssigner();
+    } else {
+      LOG.info("Iceberg source restored {} splits from state for table {}",
+          enumState.pendingSplits().size(), table.name());
+      assigner = assignerFactory.createAssigner(enumState.pendingSplits());
+    }
+
+    if (scanContext.isStreaming()) {
+      // Ideally, operatorId should be used as the threadPoolName as Flink guarantees its uniqueness within a job.
+      // SplitEnumeratorContext doesn't expose the OperatorCoordinator.Context, which would contain the OperatorID.
+      // Need to discuss with Flink community whether it is ok to expose a public API like the protected method
+      // "OperatorCoordinator.Context getCoordinatorContext()" from SourceCoordinatorContext implementation.
+      // For now, <table name>-<random UUID> is used as the unique thread pool name.
+      ContinuousSplitPlanner splitPlanner = new ContinuousSplitPlannerImpl(
+          table, scanContext, table.name() + "-" + UUID.randomUUID());
+      return new ContinuousIcebergEnumerator(enumContext, assigner, scanContext, splitPlanner, enumState);
+    } else {
+      return new StaticIcebergEnumerator(enumContext, assigner, table, scanContext, enumState);
+    }
+  }
+
+  public static <T> Builder<T> builder() {
+    return new Builder<>();
+  }
+
+  public static class Builder<T> {
+
+    // required
+    private TableLoader tableLoader;
+    private SplitAssignerFactory splitAssignerFactory;
+    private ReaderFunction<T> readerFunction;
+
+    // optional
+    private final ScanContext.Builder contextBuilder = ScanContext.builder();
+
+    Builder() {
+    }
+
+    public Builder<T> tableLoader(TableLoader loader) {
+      this.tableLoader = loader;
+      return this;
+    }
+
+    public Builder<T> assignerFactory(SplitAssignerFactory assignerFactory) {
+      this.splitAssignerFactory = assignerFactory;
+      return this;
+    }
+
+    public Builder<T> readerFunction(ReaderFunction<T> newReaderFunction) {
+      this.readerFunction = newReaderFunction;
+      return this;
+    }
+
+    public Builder caseSensitive(boolean newCaseSensitive) {
```

Comment:
@rdblue I like to revisit the discussion of exposing `ScanContext` directly to users (instead of replicating the `ScanContext` methods in the source builder here). Not only we reduced code duplication. We also avoid potential out-of-sync problem. In the past, I have seen the case where we added new methods to `ScanContext` but forgot to add them to source builder.

Reply:
@yittg @kbendick what's your take?

Reply:
I think this is making the API more complicated for the convenience of developers, so I don't think that we should expose it.